### DISCUSSION TITLE
Is there a replacement for printf in a kernel

### DISCUSSION THREAD

**Discussion by EmmanuelMess** [eyalroz/cuda-api-wrappers#389] [Q&A] [answered]
I think I missed how to print from a kernel, so I have just been using `printf` from cuda. Is there a way to use something like streams?

**Comment by eyalroz** [comment, marked as answer]
That would be a device-side facility, while this is a host-side library. I had toyed with a cout-like stream in my [cuda-kat](https://github.com/eyalroz/cuda-kat/) library, but the implementation I had was incredibly slow to compile. For now, we have my additionals for [this issue](https://github.com/eyalroz/cuda-kat/issues/98), which provides more of the C standard library's printf functionality, but can also serve as a basis for a stream-like facility; or better yet, an [std::format](https://en.cppreference.com/w/cpp/utility/format/format) / [fmt::format](https://github.com/fmtlib/fmt)-like function.

Are you sure you like the former rather than the latter?

**Reply by EmmanuelMess** [reply]
I would be happy with something similar to "ROS_*_STREAM" functions, those are used like this: `ROS_ERROR_STREAM("bla" << 8)`

**Reply by eyalroz** [reply]
@EmmanuelMess : Again, that would be a cuda-kat request. But I'm not familiar with these "ROS" functions; plus, that looks like macro-use, which I try to avoid.
